### PR TITLE
feat: Add `--watch-exclude` flag

### PR DESCRIPTION
**Pull request by lczerniawski**
This PR introduces the feature mentioned in #19734 

It introduces the ability to exclude certain paths from the file watcher in Deno. This is particularly useful when running scripts in watch mode, as it allows developers to prevent unnecessary restarts when changes are made to files that do not affect the running script, or when executing scripts that generate new files which results in an infinite restart loop.

I also created follow-up PR in the documentation repo with info about the introduction of this flag.
[Deno Docs PR](https://github.com/denoland/deno-docs/pull/272)

### CODE REVIEW

**Comment by lczerniawski** [comment]
Can someone trigger CI for me, please?

**Comment by lczerniawski** [comment]
Thanks! Can we rerun them, please? I synchronised with the main, I don't think that my changes could caused the fail that happened.

**Comment by lczerniawski** [comment]
Can we trigger CI once again? I checked why CI failed but I don't think that my changes should impact this part I merged the main once more, moreover, all tests pass locally.

**Comment by lczerniawski** [comment]
Hi @dsherret, hope you don't mind tagging you here, do you mind taking a look into that in your spare time?

File: cli/util/file_watcher.rs [hunk 1]
```diff
@@ -376,28 +396,42 @@ fn new_watcher(
...
+  watcher: &mut RecommendedWatcher,
+  paths: &[PathBuf],
+  paths_to_exclude: &Arc<Mutex<Vec<PathBuf>>>,
+) {
   // Ignore any error e.g. `PathNotFound`
+  let paths_to_exclude = paths_to_exclude.lock().clone();
+  let mut watched_paths = Vec::new();
+
   for path in paths {
+    if paths_to_exclude.iter().any(|p| path.ends_with(p)) {
```
Comment (bartlomieju) [inline cli/util/file_watcher.rs:416]: Are these paths normalized in any way? Checking just `ends_with()` seems error-prone if paths are not fully resolved.

Comment (lczerniawski) [reply cli/util/file_watcher.rs]: Hi @bartlomieju, thanks for reviewing that!

Good point, they are normalized, I changed it to equality check.

**Review by dsherret — APPROVED** [review]
LGTM. Thanks @lczerniawski!

I updated this to work with the other sub commands so that it can land for 1.42 which is being released momentarily.

**Review by bartlomieju — APPROVED** [review]

**Comment by lczerniawski** [comment]
Great!

Thank you both for your help with finishing up that topic!

**Pull request merged**